File: Introductory Problems/Grid Paths.cpp
#include <iostream>
using namespace std;

int grid[7][7];
string stng = "";
int solve(string mv, int x, int y)
{
    if(x==0&&y==6)
    {
        if(mv.length()==48)
        {
            bool founder = true;
            for(int i=0;i<48;i++)
            {
                if(stng[i]!=mv[i]&&stng[i]!='?'){ founder=false; break;}
            }
            if(founder) return 1;
            else return 0;
        }
        else return 0;
    }
    int count = 0;
    if(mv.length()!=0 && !(stng[mv.length()-1]=='?' || stng[mv.length()-1]== mv[mv.length()-1])) return 0;
    if(0<x&&x<6 && 0<y&&y<6 && ((grid[y-1][x]==grid[y+1][x]&&grid[y+1][x]==0 && grid[y][x-1]==grid[y][x+1]&&grid[y][x+1]==1) || (grid[y-1][x]==grid[y+1][x]&&grid[y+1][x]==1 && grid[y][x-1]==grid[y][x+1]&&grid[y][x+1]==0))) 
        return 0;
    else if((0<x&&x<6 && grid[y][x-1]==grid[y][x+1]&&grid[y][x+1]==0 &&grid[(y==0?1:5)][x]==1/*&& grid[y-abs(y-3)/(y-3)][x]==1)*/) || (0<y&&y<6 && grid[y+1][x]==grid[y-1][x]&&grid[y-1][x]==0 && grid[y][(x==0?1:5)]==1 /*&& grid[y][x-abs(x-3)/(x-3)]==1*/)) 
        return 0;

    if(y!=6 && grid[y+1][x]!=1)
    {
        grid[y+1][x]=1;
        mv.append("D");
        count+=solve(mv, x, y+1);
        mv.erase(mv.length()-1);
        grid[y+1][x]=0;
    }
    if (y!=0 && grid[y-1][x]!=1)
    {
        grid[y-1][x]=1;
        mv.append("U");
        count+=solve(mv, x, y-1);
        mv.erase(mv.length()-1);
        grid[y-1][x]=0;
    }
    if (x!=6 && grid[y][x+1]!=1)
    {
        grid[y][x+1]=1;
        mv.append("R");
        count+=solve(mv, x+1, y);
        mv.erase(mv.length()-1);
        grid[y][x+1]=0;
    }
    if(x!=0 && grid[y][x-1]!=1)
    {
        grid[y][x-1]=1;
        mv.append("L");
        count+=solve(mv, x-1, y);
        mv.erase(mv.length()-1);
        grid[y][x-1]=0;
    }
    return count;

}


int main()
{
    for(int i=0; i<7; i++)
    {
        for(int j=0; j<7; j++)
        {
            grid[i][j]=0;
        }
    }
    grid[0][0]=1;
    cin>>stng;

    cout<<solve("", 0, 0);
    return 0;
}

File: CSES/Introductory Problems/Grid Paths.py
grid = [[0 for i in range(7)] for j in range(7)]

def solve(mv, x, y):
    if x==0 and y==6:
        if len(mv)==48:
            found = True
            for i in range(48):
                if stng[i]!=mv[i] and stng[i]!='?':
                    found = False
                    break
            if found:
                return 1
        else:
            return 0
    count = 0
    if mv!=[] and not (stng[len(mv)-1]=='?' or stng[len(mv)-1]==mv[len(mv)-1]):
        return 0
    if 0<x<6 and 0<y<6:
        if (grid[y-1][x]==grid[y+1][x]==0 and grid[y][x-1]==grid[y][x+1]==1) or (grid[y-1][x]==grid[y+1][x]==1 and grid[y][x-1]==grid[y][x+1]==0):
            return 0
    elif (0<x<6 and grid[y][x-1]==grid[y][x+1]==0 and grid[y-abs(y-3)//(y-3)][x]==1) or (0<y<6 and grid[y+1][x]==grid[y-1][x]==0 and grid[y][x-abs(x-3)//(x-3)]==1):
        return 0

    if y!=6 and grid[y+1][x]!=1:
        grid[y+1][x]=1
        mv.append('D')
        count+=solve(mv, x, y+1)
        mv.pop()
        grid[y+1][x]=0
    if y!=0 and grid[y-1][x]!=1:
        grid[y-1][x]=1
        mv.append('U')
        count+=solve(mv, x, y-1)
        mv.pop()
        grid[y-1][x]=0
    if x!=6 and grid[y][x+1]!=1:
        grid[y][x+1]=1
        mv.append('R')
        count+=solve(mv, x+1, y)
        mv.pop()
        grid[y][x+1]=0
    if x!=0 and grid[y][x-1]!=1:
        grid[y][x-1]=1
        mv.append('L')
        count+=solve(mv, x-1, y)
        mv.pop()
        grid[y][x-1]=0
    return count

grid[0][0]=1
stng = input()
print(solve([], 0, 0))